aidial_adapter_vertexai/universal_api/storage.py

```diff
@@ -0,0 +1,142 @@
+import base64
+import hashlib
+import io
+import os
+from typing import Mapping, Optional, TypedDict
+from urllib.parse import urljoin
+
+import aiohttp
+from aiohttp import hdrs
+
+from aidial_adapter_vertexai.utils.auth import Auth
+from aidial_adapter_vertexai.utils.log_config import app_logger as logger
+
+
+class FileMetadata(TypedDict):
+    name: str
+    parentPath: str
+    bucket: str
+    url: str
+    type: str
+
+    contentLength: int
+    contentType: str
+
+
+class FileStorage:
+    dial_url: str
+    upload_dir: str
+    auth: Auth
+    bucket: Optional[str]
+
+    def __init__(self, dial_url: str, upload_dir: str, auth: Auth):
+        self.dial_url = dial_url
+        self.upload_dir = upload_dir
+        self.auth = auth
+        self.bucket = None
+
+    async def _get_bucket(self, session: aiohttp.ClientSession) -> str:
+        if self.bucket is None:
+            async with session.get(
+                f"{self.dial_url}/v1/bucket",
+                headers=self.auth.headers,
+            ) as response:
+                response.raise_for_status()
+                body = await response.json()
+                self.bucket = body["bucket"]
+
+        return self.bucket
+
+    @staticmethod
+    def _to_form_data(
+        filename: str, content_type: str, content: bytes
+    ) -> aiohttp.FormData:
+        data = aiohttp.FormData()
+        data.add_field(
+            "file",
+            io.BytesIO(content),
+            filename=filename,
+            content_type=content_type,
+        )
+        return data
+
+    async def upload(
+        self, filename: str, content_type: str, content: bytes
+    ) -> FileMetadata:
+        async with aiohttp.ClientSession() as session:
+            bucket = await self._get_bucket(session)
+            data = FileStorage._to_form_data(filename, content_type, content)
+            ext = _get_extension(content_type) or ""
+            url = f"{self.dial_url}/v1/files/{bucket}/{self.upload_dir}/{filename}{ext}"
+
+            async with session.put(
+                url=url,
+                data=data,
+                headers=self.auth.headers,
+            ) as response:
+                response.raise_for_status()
+                meta = await response.json()
+                logger.debug(f"Uploaded file: url={url}, metadata={meta}")
+                return meta
+
+    async def upload_file_as_base64(
+        self, data: str, content_type: str
+    ) -> FileMetadata:
+        filename = _compute_hash_digest(data)
```

Comment:
Is it intentional that file will have same name for same data? What is expected when same data is uploaded twice?

Reply:
Yes. Store is implemented via PUT request, which is supposed to be idempotent: two identical requests must result in the identical responses.